### PR TITLE
support backend schema

### PR DESCRIPTION
**Pull request by ivywei0125**
# Description

support backed schema

### CODE REVIEW

File: sdk/spring/spring-cloud-azure-feature-management/src/main/java/com/azure/spring/cloud/feature/management/implementation/FeatureManagementProperties.java [hunk 1]
```diff
@@ -105,13 +92,57 @@ private void addToFeatures(Map<? extends String, ? extends Object> features, Str
         }
     }
 
+    private void addServerSideFeature(Map<? extends String, ? extends Object> features, String key, String combined) {
+        final Object featureValue = features.get(key);
+        if (!combined.isEmpty() && !combined.endsWith(".")) {
```
Comment (mrm9084) [inline sdk/spring/spring-cloud-azure-feature-management/src/main/java/com/azure/spring/cloud/feature/management/implementation/FeatureManagementProperties.java:97]: I don't think the "combined" thing should be an issue. The combined issue is to deal with the fact that the name is part of the key in the client schema. With the client side schema the id filed will be the name.

File: sdk/spring/spring-cloud-azure-feature-management/src/main/java/com/azure/spring/cloud/feature/management/implementation/FeatureManagementProperties.java [hunk 2]
```diff
@@ -105,13 +92,57 @@ private void addToFeatures(Map<? extends String, ? extends Object> features, Str
...
+                featureManagement.put(serverSideFeature.getId(), feature);
+            } else {
+                onOff.put(serverSideFeature.getId(), serverSideFeature.isEnabled());
+            }
+        }
+    }
+
+    // todo what is the best way to decide whether is backend schema?
+    private boolean isBackendSchema(Map<? extends String, ? extends Object> features) {
+        if (features.keySet().isEmpty()) {
```
Comment (mrm9084) [inline sdk/spring/spring-cloud-azure-feature-management/src/main/java/com/azure/spring/cloud/feature/management/implementation/FeatureManagementProperties.java:132]: It's not a great option, but we could just check if the id field exists. It shouldn't exist for the other one. Also, maybe we default to this option, if we see it and we can fall back to the other option on error.

**Review by mrm9084 — CHANGES_REQUESTED** [review]

File: sdk/spring/spring-cloud-azure-feature-management/src/test/resources/application-client.yaml [hunk 3]
```diff
@@ -0,0 +1,15 @@
+feature-management:
+    feature-flags:
```
[the remark this answers is not in the document]
Comment (ivywei0125) [reply sdk/spring/spring-cloud-azure-feature-management/src/test/resources/application-client.yaml:2]: Ok, I see. So do we also need to update the README file? 
https://github.com/Azure/azure-sdk-for-java/tree/main/sdk/spring/spring-cloud-azure-feature-management#feature-flag-declaration
It seems that there shouldn't be a `feature-flag` section and the `time-window-filter-setting-start` parameter should be `Start`.

File: sdk/spring/spring-cloud-azure-feature-management/src/main/java/com/azure/spring/cloud/feature/management/implementation/FeatureManagementProperties.java (hunk 2 above)
Comment (ivywei0125) [reply sdk/spring/spring-cloud-azure-feature-management/src/main/java/com/azure/spring/cloud/feature/management/implementation/FeatureManagementProperties.java]: That make sense. I updated the logic to try to parse the properties by server side schema at first. If the parse process failed or get no features after this parse process, then we need to try parse by client side schema.

File: sdk/spring/spring-cloud-azure-feature-management/src/main/java/com/azure/spring/cloud/feature/management/implementation/FeatureManagementProperties.java (hunk 1 above)
Comment (ivywei0125) [reply sdk/spring/spring-cloud-azure-feature-management/src/main/java/com/azure/spring/cloud/feature/management/implementation/FeatureManagementProperties.java]: For server side schema, the key should be specified by `id` feild, so we don't need to deal with combined issue.

Comment (ivywei0125) [reply sdk/spring/spring-cloud-azure-feature-management/src/main/java/com/azure/spring/cloud/feature/management/implementation/FeatureManagementProperties.java]: Do we support the following properties? I add an additional "FeatureSet" field. After parse, shall we get no feature flags or get a feature flag with key"FeatureSet.Alpha"?
```
// application.yaml
FeatureManagement:
    FeatureSet:
        FeatureFlags:
            -
                id: Alpha
                enabled: true
                conditions:
                    client_filters:
                        -
                            name: Microsoft.Random
                            parameters:
                                Value: 50
```

File: sdk/spring/spring-cloud-azure-feature-management/src/main/java/com/azure/spring/cloud/feature/management/FeatureManagementConfiguration.java [hunk 4]
```diff
@@ -19,7 +19,7 @@ class FeatureManagementConfiguration {
```
Comment (mrm9084) [inline sdk/spring/spring-cloud-azure-feature-management/src/main/java/com/azure/spring/cloud/feature/management/FeatureManagementConfiguration.java:19]: To create a feature flag that is enabled and a list of feature flags that is empty currently requires:

```yml
feature-management:
  FeatureFlags:
    -
      id: Alpha
      enabled: true
      conditions:
        client_filters: {}
```

Which is a bit weird.
Note: This is expected to return false, which it does not do.

File: sdk/spring/spring-cloud-azure-feature-management/src/main/java/com/azure/spring/cloud/feature/management/implementation/FeatureManagementProperties.java [hunk 5]
```diff
@@ -105,13 +123,37 @@ private void addToFeatures(Map<? extends String, ? extends Object> features, Str
...
+
+        ServerSideFeature serverSideFeature = null;
+        try {
+            serverSideFeature = MAPPER.convertValue(featureValue, ServerSideFeature.class);
+        } catch (IllegalArgumentException e) {
+            LOGGER.error("Found invalid feature {} with value {}.", key, featureValue.toString());
+        }
+
+        if (serverSideFeature != null && serverSideFeature.getId() != null) {
+            if (serverSideFeature.getConditions() != null && serverSideFeature.getConditions().getClientFilters() != null
```
Comment (mrm9084) [inline sdk/spring/spring-cloud-azure-feature-management/src/main/java/com/azure/spring/cloud/feature/management/implementation/FeatureManagementProperties.java:137]: `serverSideFeature.getConditions().getClientFilters().size() > 0` is allowed to be 0. This should result in a feature flag returning false when enabled, as no feature filter returned true.

This is how the current library works.

**Review by mrm9084 — CHANGES_REQUESTED** [review]

File: sdk/spring/spring-cloud-azure-feature-management/src/main/java/com/azure/spring/cloud/feature/management/FeatureManagementConfiguration.java (hunk 4 above)
Comment (ivywei0125) [reply sdk/spring/spring-cloud-azure-feature-management/src/main/java/com/azure/spring/cloud/feature/management/FeatureManagementConfiguration.java:19]: Yes. I agree that it's a nit wired. I used to think about to support both `array` and `Map<Integer, FeatureFilterEvaluationContext>` for the "client_filters" property. Would it be better?
If we support array, then we could set an empty list with the following
```
feature-management:
  FeatureFlags:
    -
      id: Alpha
      enabled: true
      conditions:
        client_filters: []
```
And also we should support the map.
```
FeatureManagement:
    FeatureFlags:
        -
            id: Alpha
            enabled: true
            conditions:
                client_filters:
                    -
                        name: Microsoft.Random
                        parameters:
                            Value: 50
                    -   
                        name: Microsoft.TimeWindowFilter
                        parameters:
                            Start: "Thu, 29 Jun 2023 07:00:00 GMT"
                            End: "Wed, 30 Aug 2023 07:00:00 GMT"
```

File: sdk/spring/spring-cloud-azure-feature-management/src/main/java/com/azure/spring/cloud/feature/management/implementation/FeatureManagementProperties.java [hunk 6]
```diff
@@ -40,41 +41,58 @@ public FeatureManagementProperties() {
...
-            addToFeatures(m, key, "");
+    private void tryServerSideSchema(Map<? extends String, ? extends Object> features) {
+        if (features.keySet().isEmpty()) {
+            return;
         }
 
+        // check if FeatureFlags section exist
+        String featureFlagsSectionKey = "";
+        for (String key : features.keySet()) {
+            if ("FeatureFlags".equalsIgnoreCase(key)) {
```
[the remark this answers is not in the document]
Comment (ivywei0125) [reply sdk/spring/spring-cloud-azure-feature-management/src/main/java/com/azure/spring/cloud/feature/management/implementation/FeatureManagementProperties.java]: Updated

File: sdk/spring/spring-cloud-azure-feature-management/src/main/java/com/azure/spring/cloud/feature/management/implementation/FeatureManagementProperties.java (hunk 5 above)
Comment (ivywei0125) [reply sdk/spring/spring-cloud-azure-feature-management/src/main/java/com/azure/spring/cloud/feature/management/implementation/FeatureManagementProperties.java]: Updated.

File: sdk/spring/spring-cloud-azure-feature-management/src/main/java/com/azure/spring/cloud/feature/management/FeatureManagementConfiguration.java (hunk 4 above)
Comment (mrm9084) [reply sdk/spring/spring-cloud-azure-feature-management/src/main/java/com/azure/spring/cloud/feature/management/FeatureManagementConfiguration.java:19]: I may be missing something, but both of those are lists for the client_filters field.

**Review by saragluna — APPROVED** [review]

File: sdk/spring/spring-cloud-azure-feature-management/src/main/java/com/azure/spring/cloud/feature/management/implementation/FeatureManagementProperties.java (hunk 6 above)
[the remark this answers is not in the document]
Comment (ivywei0125) [reply sdk/spring/spring-cloud-azure-feature-management/src/main/java/com/azure/spring/cloud/feature/management/implementation/FeatureManagementProperties.java]: Of course.

File: sdk/spring/spring-cloud-azure-feature-management/src/main/java/com/azure/spring/cloud/feature/management/FeatureManagementConfiguration.java (hunk 4 above)
Comment (mrm9084) [reply sdk/spring/spring-cloud-azure-feature-management/src/main/java/com/azure/spring/cloud/feature/management/FeatureManagementConfiguration.java:19]: If it's no a new dependency to the SKD library that I'm good with it.

**Review by mrm9084 — APPROVED** [review]

**Pull request merged**